smack-experimental/src/main/java/org/jivesoftware/smackx/chat_markers/ChatMarkersManager.java

```diff
@@ -87,4 +205,101 @@ public boolean isSupportedByServer()
                 .serverSupportsFeature(ChatMarkersElements.NAMESPACE);
     }
 
+    /**
+     * Register a ChatMarkersListener. That listener will be informed about new
+     * incoming markable messages.
+     *
+     * @param listener ChatMarkersListener
+     * @return true, if the listener was not registered before
+     */
+    public boolean addIncomingChatMarkerMessageListener(ChatMarkersListener listener) {
+        synchronized (incomingListeners) {
+            return incomingListeners.add(listener);
+        }
+    }
+
+    /**
+     * Unregister a ChatMarkersListener.
+     *
+     * @param listener ChatMarkersListener
+     * @return true, if the listener was registered before
+     */
+    public boolean removeIncomingChatMarkerMessageListener(ChatMarkersListener listener) {
+        synchronized (incomingListeners) {
+            return incomingListeners.remove(listener);
+        }
+    }
+
+    /**
+     * From XEP-0333, Protocol Format: The Chat Marker MUST have an 'id' which is the 'id' of the
+     * message being marked.<br>
+     * In order to make Chat Markers works together with XEP-0085 as it said in
+     * 8.5 Interaction with Chat States, only messages with <tt>active</tt> chat
+     * state are accepted.
+     *
+     * @param message to be analyzed.
+     * @return true if the message contains a stanza Id.
+     * @see <a href="http://xmpp.org/extensions/xep-0333.html">XEP-0333: Chat Markers</a>
+     */
+    private boolean shouldDiscardMessage(Message message) {
+        if (StringUtils.isNullOrEmpty(message.getStanzaId())) {
+            return true;
+        }
+
+        if (!CHAT_STATE_FILTER.accept(message)) {
+            ExtensionElement extension = message.getExtension(ChatStateManager.NAMESPACE);
+            String chatStateElementName = extension.getElementName();
+
+            ChatState state;
+            try {
+                state = ChatState.valueOf(chatStateElementName);
+                return !(state == ChatState.active);
+            }
+            catch (Exception ex) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    public void markMessageAsReceived(String id, Jid to, Jid from, String thread)
```

Comment:
Maybe you could instead offer a method `markMessageAsReceived(Message message)`. That would make more sense from a users perspective I guess. Same goes for the other `markMessageAsXYZ` methods.

Reply:
Going the way you propose would introduce either more code to validate the message structure is the correct one or the possibility to get some unpredicted behavior. i.e.: if the message has a composing tag, or another optional tags that XEP-0333 doesn't allow.
But you know this project better than me, so I will do what do  you think is better :)

Reply:
Ping @Flowdalic

Reply:
My first guess would be, that the methods should validate, that the message has the correct form and if not throw respective exceptions. Currently the user is able to input invalid too if I understand the code correctly.

Reply:
sorry for the late answer, I will update this as you suggest.

Reply:
Done, can you give me a feed back?